Feature/semi step1

semi-cloud opened the pull request:
[파일 저장소 선정]
- DB 저장 VS 파일 서버에 저장
- 데이터베이스에 바이너리 데이터로 저장하는 것은 대용량 데이터일 경우 쓰기 연산으로 읽기 연산 성능까지 해치는 경우가 발생할 것 같아, 따로 파일 서버를 두고 저장하는 방식을 택했습니다.

[FileService]
- 파일 업로드 로직 관련 고민

>1-1. 사용자가 파일을 업로드한다. -> 서버에 파일이 저장
1-2. 파일 메타데이터 DB 저장 과정에서 예외 발생 -> 롤백 및 사용자에게 파일 업로드 실패 에러 반환
1-3. 해당 메시지를 본 사용자가 다시 파일 업로드 -> 중복 파일 로직으로 간주가 되지 않아 파일이 중복으로 저장

해당 과정을 방지하기 위해, 파일 메타데이터를 저장하는 과정을 파일을 저장하는 과정보다 앞으로 이동시켰습니다.
>1-1. 파일 메타데이터 DB 저장 과정에서 예외 발생 -> 롤백 및 사용자에게 파일 업로드 실패 에러 반환
1-2. 해당 메시지를 본 사용자가 다시 파일 업로드 -> 중복 파일 로직으로 간주가 되지 않아 파일 메타데이터 다시 저장 시도
1-3. 성공적으로 메타데이터가 업로드 된다면 파일을 업로드

- 파일 업로드 시, 만약 악성 유저가 동시성 문제를 일으킬 만큼 빠르게 업로드를 누르게 된다면 `isDuplicateFile` 검증 부분에서 `ReadLock` 을 안걸었기 때문에 중복으로 파일이 저장되는 상황이 생길 것이라 생각했습니다.(혹시 잘못된 지식이라면..🥹) 하지만 현실적으로 불가능하다 생각되어 해당 부분은 처리를 하지 않았습니다.

[FileController]
- `@NotBlank`는 문자열 대상이라 `MultipartFile` 과 같은 객체에는 적용되지 않았습니다. 따라서 따로 서비스단에서 `null` 및 파일 내용이 비어있을 때 처리를 해주었습니다. (다른 분들도 테스트해서 확인해보셨으면 좋겠어요) 
- 한가지 궁금한 점이 `@Notnull` 조건을 위배하면 왜 바인드 관련 예외가 안터지고  `MethodArgumentTypeMismatchException` 이 터지는지 좀 더 찾아봐야 할 것 같습니다. 일단 해당 이유 때문에 `CustomExceptionHandler` 에서 바인드 관련 두 예외에 더불어 해당 예외까지 `4XX` 에러를 반환하도록 처리해주었습니다.
- 저 같은 경우 `@Validated` 를 붙이지 않으면 `@Valid` 가 동작하지 않던데, 다른 분들은 성공하셨는지 궁금하네요.

[CustomExceptionHandler]
- 예외 발생하면 스택 트레이스 보면서 해결하는게 편해서 스택 트레이스 전체를 남기도록 했습니다. 

[Etc]
- 자바 17에서는 코틀린처럼 `val(final)`을 지원하더라구요 대박,, 그래서 타입 안쓰고 죄다 `val` 로 바꿨는데 오히려 가독성 측면에서는 떨어지려나? 🤔

semi-cloud committed:
feat: 데이터베이스 초기 세팅

Commit:
chore: update checkstyle version

semi-cloud committed:
feat: 파일 업로드 기능 구현

[FileController]
- @Validated를 붙이지 않으면 @Valid가 동작하지 않아서 붙임
- 테스트도 가능하도록 생성한 데이터를 같이 반환

[FileService]
- 파일이 null 이거나 내용이 비어 있는 경우 예외 처리
- 사용자가 실수록, 혹은 메타데이터 저장에서 예외가 발생해서 트랜잭션이 롤백될 때 다시 사용자가 파일을 추가하게끔 하더라도 같은 파일이 저장되지 않도록 파일 중복 예외 처리

[FileUtil]
- 비즈니스 예외와 관계 없는 I/O 체크 예외를 분리해내기 위해 생성
- 서비스 테스트에서 파일 테스트에 대한 의존성까지 분리 가능

[CustomExceptionHandler]
- @Notnull 조건 위반시 bind와 관련 없는 MethodArgumentTypeMismatchException이 발생해서 같이 핸들링 하도록 조건에 추가

[FileServiceTest]
- 파일 중복 테스트를 위해 @SpringBootTest 도입, 메모리 DB 연결

semi-cloud committed:
feat: 파일 다운로드 기능 구현

[FileController]
- 파일 타입에 맞춰서 보여줄 수 있도록 바이트 배열로 반환
- 사용자 지정 파일 다운로드 위치는 고려하지 않아도 되는건지 의문

[FileService]
- 본인이 업로드한 파일이 아닐 경우 접근 권한 예외 처리
- DTO 변환 로직에서 체크 예외가 발생하는데 DTO 내부로 둬야할지, 서비스 레이어로 뺄지 고민

[FileUtil]
- 리소스를 경로에서 가져왔으나 비어있거나, 읽지 못하는 경우 예외 처리

[FileServiceTest]
- @SpringBootTest 제거, Mocking으로 의존성 분리

semi-cloud committed:
test: Controller, FileUtil 테스트 추가

[FileUtilTest]
- UrlResource mocking에 실패해 @Test 주석

semi-cloud committed:
feat: 파일 삭제 기능 구현

[FileUtil]
- mockstatic + logger 조합일 때 예외 발생으로 인해 제거

semi-cloud committed:
refactor: 파일 다운로드 로직 수정

[FileUtil]
- new UrlResource를 Mocking 불가능하므로 ResourceLoader로 수정
- ResourecLoader @Mock을 하면서 Util에 관련 기능을 모으기 위해 static 제거

semi-cloud committed:
Merge branch 'main' into feature/semi_step1

semi-cloud committed:
test: 파일 다운로드 서비스 예외 테스트 추가

- 테스트 버그 수정
- 다운로드 파일 바이트 배열로 변환하는 과정에서 발생하는 예외 테스트

semi-cloud committed:
refactor: 파일 업로드 로직 수정

- 롤백 시 파일이 중복으로 저장되는 것을 방지하기 위해 메타데이터를 먼저 업로드 하도록 변경

semi-cloud committed:
style: 공백 수정

semi-cloud committed:
refactor: 코드 스멜 및 버그 개선

[FileDownloadDto]
- equals/hashcode 위해 byte[] 필드 Record 클래스로 분리

[BusinessException]
- debugMessage 필드 final 추가

[FileService]
- 문자열 연산으로 생성된 URL에서 보안상 취약이 발생하는 것을 막기 위해 검증 로직 추가
- 타입을 붙이는 것이 코드를 이해하기 쉽다 생각해 val 제거
- @SpringBootTest없이 @Value 값을 넣기에 불편한 점이 많아 상수 필드로 수정

semi-cloud committed:
refactor: equals, hashcode 구현 및 상수 필드명 수정

VSFe commented on src/main/java/com/c4cometrue/mystorage/api/dto/FileDownloadDto.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.util.Arrays;
+
+public class FileDownloadDto {
```
DTO 라고 하면, 저희는 이 클래스를 그대로 반환하길 기대할 것 같은데, 다른 구조로 되어 있는 것 같습니다.

VSFe commented on src/main/java/com/c4cometrue/mystorage/common/config/JpaConfig.java:
```diff
@@ -0,0 +1,9 @@
+
+@EnableJpaAuditing
+@Configuration
+public class JpaConfig {
```
아무 메서드도 들어있지 않으면, 굳이 Config 클래스를 만들 필요는 없어보입니다.

`@EnableJpaAuditing`은 Main Class로 옮겨주시고, 해당 파일을 삭제해주세요.

semi-cloud replied:
`@SpringBootApplication`와 같이 위치시켰을 때 해당 메인 클래스의 어노테이션들을 모두 불러오게 되는데, `@WebMvcTest`에서는 `@EnableJpaAuditing` 에 필요한 `JPA` 관련 클래스들을 불러오지 않으니 아래와 같은 에러가 나서 분리시켜주었습니다.
 
매 컨트롤러 테스트 클래스마다 `@MockBean(JpaMetamodelMappingContext.class)` 을 입력하는 방식이 나을까요?

```
Error creating bean with name 'jpaMappingContext': JPA metamodel must not be empty
```

VSFe commented on src/main/java/com/c4cometrue/mystorage/common/exception/ErrorCode.java:
```diff
@@ -0,0 +1,33 @@
+	// common
+	INTERNAL_SERVER_ERROR(HttpStatus.INTERNAL_SERVER_ERROR, "서버 오류입니다."),
+	FILE_NOT_FOUND(HttpStatus.NOT_FOUND, "파일이 존재하지 않습니다."),
+	INVALID_FILE_PATH(HttpStatus.NOT_FOUND, "올바른 파일 경로가 아닙니다."),
```
`NOT_FOUND`가 아니라, `BAD_REQUEST`가 더 맞는 것 같아요. (사용자 입력값의 문제)

MilanoBeer reviewed (commented):
수고하셧슴당

VSFe commented on src/main/java/com/c4cometrue/mystorage/common/exception/handler/CustomExceptionHandler.java:
```diff
@@ -0,0 +1,45 @@
+		return ResponseEntity.internalServerError().body(ErrorCode.INTERNAL_SERVER_ERROR.getMessage());
+	}
+
+	private boolean isValidException(Exception ex) {
```
모든 예외를 이렇게 처리하는건 불가능하겠죠? (이후에도 다양한 작업을 할 테니)
결국 우리는 CheckedException, UncheckedException으로 나눠서 전략을 세울 필요가 있습니다.

- CheckedException -> throw 하거나, catch 해야하는데, 발생해야 하면 catch 해서 BusinessException으로 Wrapping
- UncheckedExcpetion -> 어차피 RuntimeException으로 나오니, INTERNAL_SERVER_ERROR 반환하게 처리

추가적으로, Wrapping한 예외가 아닌 이상, 절대로 외부에 에러 메시지를 노출하게 하지 마세요. (왜냐고요? 직접 띄워보세요!)

semi-cloud replied:
> @NotNull(message = "유저 ID는 null이 될 수 없습니다.")

여기서 이 `message` 를 클라이언트로 반환해줘야 할 것 같은데, `Exception ex` 로 바인드 예외까지 모두 잡아버리면 `INTERNAL_SERVER_ERROR` 에러 코드 메시지인 `서버 오류입니다` 라고만 나오더라구요 🥹

이런 경우는 어떻게 처리하는게 좋을 지 여쭤보고 싶습니다!

VSFe commented on src/main/java/com/c4cometrue/mystorage/domain/FileMetaData.java:
```diff
@@ -0,0 +1,46 @@
+	@Column(nullable = false)
+	private long size;
+
+	@Column(nullable = false)
```
type은 enum으로 하는게 관리하기 쉬울 것 같아요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,146 @@
+@RequiredArgsConstructor
+public class FileService {
+
+	private static final String UPLOAD_DIR = "/Users/kangsemi/Desktop/git/files/";
```
디렉토리는 가능하면 `@Value`로 주입받을 수 있도록 해 주세요.

semi-cloud replied:
아 이거는 청천님한테도 말씀드렸던 것처럼 테스트 때문에 바꾼거라!
다시 수정해놨었는데 왜 이전 버전으로 보이지? 😵

VSFe commented on src/main/java/com/c4cometrue/mystorage/common/exception/ErrorCode.java:
```diff
@@ -0,0 +1,33 @@
+	DUPLICATE_FILE(HttpStatus.BAD_REQUEST, "이미 존재하는 파일입니다."),
+
+	// file-download
+	INVALID_FILE_ACCESS(HttpStatus.UNAUTHORIZED, "다운로드할 파일에 접근할 권한이 없습니다."),
```
https://developer.mozilla.org/ko/docs/Web/HTTP/Status/401 
참고해주세요.

semi-cloud replied:
우하하 `403` 이군요!! 이런 실수를..

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,146 @@
+		return new FileUploadDto.Response(fileMetaData);
+	}
+
+	private boolean isDuplicateFile(String fileName, long userId) {
```
public으로 노출되는 메서드를 위쪽에 배치하고, (즉, 사용하는 메인 비즈니스 로직을 상단에 올려놓고)
private으로 노출되는 메서드를 아래쪽에 배치하면, (즉, 내부적으로 사용하는 로직은 하단에 배치하고)
코드를 사용하기 위해 확인하는 사람의 입장에서 좀 더 나을 것 같습니다.

semi-cloud replied:
으음 무작정 클린코드에 맞추는게 답은 아니군요 감사합니다
http://amazingguni.github.io/blog/2016/06/Clean-Code-10-클래스

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,146 @@
+			throw new BusinessException(ErrorCode.INVALID_FILE_ACCESS);
+		}
+
+		fileUtil.deleteFile(fileMetaData.getUploadName());
```
데이터베이스에서 파일 삭제가 이뤄지지 않네요.

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/api/dto/FileDownloadDto.java:
```diff
@@ -0,0 +1,40 @@
+
+public class FileDownloadDto {
+
+	public record Response(
```
DTO만드실때 클래스안에 클래스를 넣은 이유가 있으실까요!

kochungcheon replied:
nested 클래스에 static 을 안쓰신 이유가 있으신가요!?

kochungcheon replied:
레코드만 쓰시면 밑의 오버라이드 안해도 되요 레코드가 다해주거든요!

semi-cloud replied:
> DTO만드실때 클래스안에 클래스를 넣은 이유가 있으실까요!

한 Dto 안에 Request - Response 넣어서 하면 알아보기가 좀 더 편하더라구요 DTO 개수도 줄고?
근데 이건 Request를 @Valid 테스트하면서 빼버리는 바람에 Response만 달랑 남아서 모양새가 좀 그렇네요 하하

> nested 클래스에 static 을 안쓰신 이유가 있으신가요!

record 클래스라면 static을 중복으로 써줄 필요가 없습니다

<img width="952" alt="image" src="https://github.com/C4-ComeTrue/my-storage/assets/71436576/8f8fe65a-f693-4523-92f0-36088a97ca3b">

[참고 링크](https://docs.oracle.com/en/java/javase/15/language/records.html#GUID-6699E26F-4A9B-4393-A08B-1E47D4B2D263)


> 레코드만 쓰시면 밑의 오버라이드 안해도 되요 레코드가 다해주거든요!

네 맞아요 레코드로 바꿨었는데 쏘나큐브가 계속 채찍질을,, 그래서 어쩔 수 없이 오버라이딩 해주었습니다

kochungcheon replied:
오 그렇군요! 배우고갑니다!

MilanoBeer commented on src/main/java/com/c4cometrue/mystorage/common/exception/ErrorCode.java:
잘못된 파일경로를 입력해서 에러를 반환할 수 있는 경우는 생각을 못해봤었는데, 보고 배웠습니당!

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/api/FileController.java:
```diff
@@ -0,0 +1,61 @@
+	}
+
+	@GetMapping
+	public ResponseEntity<byte[]> download(
```
반환값으로 byte[]를 하신 이유가 궁금합니다!

semi-cloud replied:
사용자가 다운로드 버튼 클릭하면 이렇게 창에 띄우려고 바이트로 반환하고 `ContentType` 지정해주었습니다!

<img width="1010" alt="image" src="https://github.com/C4-ComeTrue/my-storage/assets/71436576/22de5817-a752-438d-9b0a-7f15034b0b1b">

semi-cloud committed:
refactor: 파일 경로 보안 문제 해결

[FileUtil]
- normalize 메서드를 통해 경로의 중복되는 요소 모두 제거

semi-cloud commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,146 @@
+@RequiredArgsConstructor
+public class FileService {
+
+	private static final String UPLOAD_DIR = "/Users/kangsemi/Desktop/git/files/";
```
으음 미리 선수치자면.. `validateFileUploadPath` 테스트 검증 때문에 설정 파일에서 @Value로 주입받기 -> 상수 필드로 변경했습니다. `Path` 관련 보안 문제 해결되면 다시 원상복구할 예정!

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/repository/FileMetaDataRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface FileMetaDataRepository extends JpaRepository<FileMetaData, Long> {
+
+    boolean existsByFileNameAndUserId(String fileName, long userId);
```
쿼리가 어떻게 나가나요!?

semi-cloud replied:
`limit` 붙여서 나갑니당. 아 이거 인덱스를 걸어야할 것 같은 느낌이,,
```sql
select
        f1_0.file_meta_data_id 
    from
        file_meta_data f1_0 
    where
        f1_0.file_name=? 
        and f1_0.user_id=? limit ?
```

MilanoBeer commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
각 서비스 기능을 구현하기 위해서 정의한 다른 메소드들을 바로 하단에 두는게 좋을지, 맨 밑에 다같이 모아두는게 가독성에 좋을지,, 아니면 그다지 상관이 없을지 궁금해졌습니당

semi-cloud replied:
`private` 메서드 말씀하시는 건가용? 음 본인이 호출된 함수 바로 밑에 두는게 컨벤션이라고 알고 있어요!
<img width="872" alt="image" src="https://github.com/C4-ComeTrue/my-storage/assets/71436576/63e885df-35d0-4ad2-8796-23f2a14d759d">

[원본 링크](http://amazingguni.github.io/blog/2016/06/Clean-Code-10-클래스)

MilanoBeer replied:
오 몰랐어요.. 개인 스타일에 따른거일려나 싶었는데 일반적인 컨벤션이 있구나 배워가욥~

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,146 @@
+@RequiredArgsConstructor
+public class FileService {
+
+	private static final String UPLOAD_DIR = "/Users/kangsemi/Desktop/git/files/";
```
환경변수 따로 관리하는건 어떤가요

semi-cloud replied:
이거는 제거했으니 패스~~

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,146 @@
+
+	private void validateFileUploadPath(String path) {
+		try {
+			File file = new File(path);
```
Files 는 어떤가요!?

semi-cloud replied:
이거도 제거했으니 패스하고 싶다만 `Canonical path` 를 얻기 위해 `File` 로 생성했습니다:)

semi-cloud committed:
refactor: 불필요한 파일 경로 검증 메서드 제거

MilanoBeer commented on src/main/java/com/c4cometrue/mystorage/api/dto/FileDownloadDto.java:
```diff
@@ -0,0 +1,40 @@
+
+import java.util.Arrays;
+
+public class FileDownloadDto {
```
파일 내부를 보지않고, 다른 사람이 이름만 봤을 때 request/response를 위한 dto인지 먼저 알 수 있도록 파일명에 같이 써주면 어떨까 싶은 생각을 했습니당

semi-cloud replied:
아하 `FileDownLoadResponse` / `FileDownLoadRequest` 가 가독성에 더 좋다는 의견이시군요!
저는 위 방식처럼 하면 너무 DTO 개수가 많아져서, 같은 성질끼리 묶어서(?) 사용하는걸 선호했는데 혜연님 말씀도 맞는 것 같네요~

pear96 replied:
저도 묶어서 한적 있는데 요청과 응답이 같은 데이터를 반환하는게 아니라면 결국 따로 만들어야하고, 그러면 통일성이 깨져서 그냥 나누게 되더라구요!
그리고 이건 진짜 취향이지만 Res / Req로 줄여도 괜찮은 것 같습니다

MilanoBeer replied:
Res / Req 좋은것 같아요~ 다음에 그렇게 해봐야겠어용

MilanoBeer commented on src/main/java/com/c4cometrue/mystorage/api/FileController.java:
```diff
@@ -0,0 +1,61 @@
+	@ResponseStatus(HttpStatus.CREATED)
+	public FileUploadDto.Response upload(
+		@RequestPart MultipartFile file,
+		@NotNull(message = "유저 ID는 null이 될 수 없습니다.") @Valid long userId
```
저는 요청파라미터로 userId 대신, userName을 사용했습니다.

이 프로젝트는 회원가입 기능, user테이블이 따로 없고,
사용자는 자신이 정한 String타입의 이름을 입력할 것이기 때문에 userName을 그대로 사용했습니다.

semi-cloud replied:
아하 `userName` 을 받는 형식도, 아 이 프로젝트에는 따로 유저 기능이 없구나라고 바로 알 수 있다는 장점도 있을 것 같네요:)
저는 유저 관련 기능 및 테이블이 가상으로 있다고 생각하고 만든거라, 제가 회원 기능을 간단하게라도 추가하지 않는 이상 코드만 보고 착각할 여지가 있을 것 같습니다.

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,146 @@
+	public FileDownloadDto.Response fileDownLoad(long userId, long fileId) {
+		FileMetaData fileMetaData = getFileMetaData(fileId);
+
+		if (hasFileAccess(userId, fileMetaData.getUserId())) {
```
hasFileAccess 메서드 내부에서 에러 처리하는건 어떤가요

kochungcheon replied:
파일에 접근 할 수 있는지 메서드에 물어봤을때 메서드가 그것에 대한 검증 메시지를 던지는 게 어떨까란 생각 +

fileDelete에도 해당 메서드가 사용되는데 반복적으로 사용되는 if throw 코드를 줄일수 있어서요!

semi-cloud replied:
> hasFileAccess 메서드 내부에서 에러 처리하는건 어떤가요

아하 그러면 중복 코드를 줄일 수 있겠네요. 그렇게 되면 메서드명으로는 `has` 를 못쓰니 `validateFileAccess` 가 되어야 할 것 같구요.
```java
private void validateFileAccess(long userId, long fileUserId) {
      if (userId != fileUserId) {
          throw new BusinessException(ErrorCode.INVALID_FILE_ACCESS);
      }
}
```

kochungcheon commented on src/main/java/com/c4cometrue/mystorage/utils/FileUtil.java:
```diff
@@ -0,0 +1,59 @@
+
+@Component
+@RequiredArgsConstructor
+public class FileUtil {
```
private 생성자 만드는게 어떤가요 
util 클래스는 객체 생성이 필요 없으니깐요!

semi-cloud replied:
으잉? 만들었었는데 언제 지워졌지 👀 감사합니닷

semi-cloud replied:
ㅎㅎ 이거는 좀 많은 시행착오가 있었습니다

1. 기존에 `new UrlResource`가 `mocking` 이 불가능
2. 따라서 `ResourceLoader`를 어쩔 수 없이 외부에서 주입받아서 해결
3. `FileService` 에서 `ResourceLoader` 를 주입받는 방법도 있겠지만 파일 관련 코드들은 한곳에 모아놓고 싶었습니다.(유틸 클래스 용도에 벗어나게 되더라도)

따라서 `private` 생성자가 위의 상황에서는 불가능합니다. 잘 보시면 메서드에 `static` 도 빠져있는 것을,,볼 수가 있어요
이렇게 보니 더 이상 유틸 클래스라 부르기 좀 그렇네요 클래스명을 바꿔야 하나 🤔

pear96 replied:
저도 리소스 로더 파일 유틸에 넣으려다가 그냥 서비스가 되는 것 같아서 고민하다가 파일 서비스에 넣었어요 (이렇게 하라는게 아니라 그냥 그랬다는 뜻)

kochungcheon reviewed (commented):
고생하셨습니다!

MilanoBeer reviewed (commented):
![image](https://github.com/C4-ComeTrue/my-storage/assets/86315623/40d59748-ffe8-4c8d-bf22-417ad3642c03)

step1 수고하셨습니다! 멋져멋져

semi-cloud commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,130 @@
+				new FileDownloadDto.Bytes(file.getContentAsByteArray()),
+				fileMetaData.getType()
+			);
+		} catch (IOException ex) {
```
그리고 이거 의견 여쭤보고 싶었는데,
DTO 변환할 때 발생하는 체크 예외를 지금처럼 서비스 바깥에 빼두느냐, 아니면 DTO 내부에다가 감추느냐 고민을 했었거든요.
다들 어떻게 생각하실까요? 😮

pear96 replied:
제 생각에 예외처리는 DTO보다는 서비스 레이어에서 책임지는게 좋을 것 같습니다

semi-cloud committed:
fix: @Value 실수 처리

pear96 commented on src/main/java/com/c4cometrue/mystorage/api/FileController.java:
```diff
@@ -0,0 +1,61 @@
+	@ResponseStatus(HttpStatus.CREATED)
+	public FileUploadDto.Response upload(
+		@RequestPart MultipartFile file,
+		@NotNull(message = "유저 ID는 null이 될 수 없습니다.") @Valid long userId
```
컨트롤러에 @Validated가 있으므로 유효성검사가 수행됩니다. @Valid는 특히 복합 객체 (커스텀 클래스)에 사용한다고 생각하시면 좋을 것 같아요.
즉 커스텀 클래스 안에서 필드들의 유효성 조건을 설정해두고, 컨트롤러의 메서드에서는 @Valid 어노테이션을 활용하고, 단일 타입 (현재 long userId)와 같은 파라미터에는 @NotNull만 적용해두면 될 것 같습니다.

https://unluckyjung.github.io/spring/2022/07/03/Valid-vs-Validated/
https://www.baeldung.com/spring-valid-vs-validated

semi-cloud replied:
`@RequestParam` 은 `validation` 과정을 아예 거치지 않더라구요! 그래서 `MethodArgumentTypeMismatchException` 이 발생했던 것 같습니다

pear96 commented on src/main/java/com/c4cometrue/mystorage/common/exception/BusinessException.java:
```diff
@@ -0,0 +1,37 @@
+@Getter
+public class BusinessException extends RuntimeException {
+
+	static final long serialVersionUID = -8294120478345891911L;
```
serialVersionUID는 무슨 값 넣으라는 규칙이 있나요?
그리고 Exception도 직렬화, 역직렬화 해야하는 필요성이 뭔지 궁금합니다.
혹시 모를 예상치 못한 직렬화/역직렬화를 방지하기 위해..??

semi-cloud replied:
음 규칙은 따로 없는걸로 알고 있어요..! 

예외는 어디에서나 사용할 수 있도록 고안되었기 때문에(ex) 분산 응용 프로그램(RMI, Spring http-invoker 등 사용), 원격 클라이언트에서 호출되는 서버 메서드에서 가져올 수 있는 경우 등) 클라이언트로 이동하려면 예외를 직렬화해야 한다고 해요.
그리고 부모인 Throwable 역시 Serializable를 구현하고 있어서 UID를 붙여주었습니다. 

<img width="471" alt="image" src="https://github.com/C4-ComeTrue/my-storage/assets/71436576/ccd3f603-cb1e-4cab-9a3e-757f0eac072e">

https://softwareengineering.stackexchange.com/questions/319199/when-we-need-to-serialize-an-exception
https://stackoverflow.com/questions/7683739/why-my-exception-class-needs-to-be-serialized

pear96 commented on src/main/java/com/c4cometrue/mystorage/common/exception/handler/CustomExceptionHandler.java:
```diff
@@ -0,0 +1,45 @@
+		return ResponseEntity.internalServerError().body(ErrorCode.INTERNAL_SERVER_ERROR.getMessage());
+	}
+
+	private boolean isValidException(Exception ex) {
```
오 이렇게 예외를 분리할 수도 있겠군요.

pear96 commented on src/main/java/com/c4cometrue/mystorage/domain/BaseEntity.java:
```diff
@@ -0,0 +1,26 @@
+@EntityListeners(AuditingEntityListener.class)
+@MappedSuperclass
+@Getter
+public class BaseEntity {
```
폴더 엔티티도 만들거 생각하면 괜찮은 것 같네요

MilanoBeer replied:
저도 나중에 폴더 엔티티 추가되면 거기도 사용할 수 있겠다 싶어서 BaseEntity 따로 뺐습니닷!

pear96 commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,130 @@
+	 */
+	@Transactional
+	public FileUploadDto.Response fileUpload(MultipartFile file, long userId) {
+		if (Objects.isNull(file) || file.isEmpty()) {
```
Multipartfile이 제출되지 않은 경우 컨트롤러의 requestPart의 required=true 때문에 MssingServletRequestException이 발생됩니다! 그래서 isEmpty는 체크하는게 맞는데 null은 체크하지 않아도 되긴 해용.. 하지만 지우라는 뜻은 아닙니다.

pear96 commented:
근데 청천님이랑 세미님 enhancement 라벨 단 이유가 본인이 새로운 기능을 만들었기 때문인건가요?

VSFe commented on src/main/java/com/c4cometrue/mystorage/service/FileService.java:
```diff
@@ -0,0 +1,130 @@
+import lombok.RequiredArgsConstructor;
+
+@Service
+@Transactional(readOnly = true)
```
- `@Transactional(readOnly = true)` 의 정확한 의미는 무엇인가요?
- readOnly를 붙임으로써 어떤 변화가 발생하나요?
- 해당 어노테이션을 최상단에 배치하는게 적절할까요?

semi-cloud replied:
읽기 전용 메서드임을 알리는 `readOnly` 를 붙이므로써 발생하는 변화는 아래와 같습니다.

+ readOnly 옵션이 설정되면, JPA에서 flush 를 수동 설정으로 바꿔버리기 때문에 커밋 시 실제 쿼리가 나가는것을 막아 조회만 하는 경우에 데이터베이스 값이 변경되지 않도록 보장할 수 있습니다. 또한, 변경감지를 위한 스냅샷을 생성하지 않아 메모리 공간도 절약할 수 있습니다.

> 해당 어노테이션을 최상단에 배치하는게 적절할까요?

+ 음 반대로 최상단에 @Transactional을 배치하고 조회용 메서드에만 readOnly 옵션을 설정하는게 가독성이 나을것 같네요.

VSFe commented on src/main/java/com/c4cometrue/mystorage/common/exception/BusinessException.java:
```diff
@@ -0,0 +1,37 @@
+	public BusinessException(ErrorCode errorCode) {
+		super(errorCode.getMessage());
+		this.errorCode = errorCode;
+		this.debugMessage = null;
```
debugMessage가 null일 경우, AdviceHandler에서 어떻게 debugMessage를 처리하는지 따라가보세요. 해당 처리 방식이 과연 올바른 처리일까요?

VSFe reviewed (commented):
하하

semi-cloud committed:
refactor: 피드백 반영

[FileController]
- long 이면 null 타입을 가질 수 없으므로 @Valid 및 @NotNull 제거
- @RequestParam을 제외한 @RequestBody는 validation 과정을 거치므로 @Valid 유지 및 DTO로 빼기

[FileService]
- Objects.isNull(file)이 없으면 파일을 보내지 않았을 때 @RequestPart가 예외를 잡지 못해서 유지
- 파일 접근 권한을 확인하고 예외 발생하는 과정 메서드 하나로 묶기(중복 코드 제거)
- 가독성을 위해 @Transactional 어노테이션 상단에 배치

[CustomExceptionHandler]
- debugmessage 가 null 일 경우 핸들링 추가
- 바인드 관련 예외는 if 문으로 판단하지 않고 따로 메서드로 분리, 컨트롤러에서 발생한 예외 정보 담아서 반환

[ErrorResponse]
- 에러 메시지와 함께 컨트롤러 바인드 예외 정보 담기 위해 생성

semi-cloud commented:
@VSFe 
피드백 반영했습니다! 
피드백 사항 커밋에 빠트린게 있어서 아래 추가로 적어봅니다.  

[FileService]
- `public/private` 메서드 위치 변경
-  파일 메타데이터 삭제 로직 추가

[ErrorCode]
- `403` 으로 수정

> DTO 라고 하면, 저희는 이 클래스를 그대로 반환하길 기대할 것 같은데, 다른 구조로 되어 있는 것 같습니다.
- 추가로 DTO 쪽은 어떻게 바꿔야 좋을지 모르겠어서 일단 패스했습니다 🥹

semi-cloud committed:
refactor: 코드 스멜 제거
